### ISSUE TITLE
Nondeterminism

### ISSUE THREAD

**Issue by BebeSparkelSparkel** [nomeata/ghc-justdoit#3]
It seems like there should be an error for when a type signature is nondeterministic.

For example
```haskell
ifVal :: (a -> Bool) -> b -> b -> a -> b
ifVal = (…)
```

How could anyone from the above type signature determine the relationships between `a -> Bool`, `a`, and `b`?

**Comment by nomeata** [comment]
Yes, indeed this library just picks any solution. There are more complex algorithms that try to find a unique best solution, but I think they are out of scope for this proof of concept. If someone wants to take this over and make it more useful, I'd support that.

**Comment by BebeSparkelSparkel** [comment]
Just talking with @TOTBWF about this. I'm not sure if I'm capable of contributing to this yet. Is there a todo list?

**Comment by BebeSparkelSparkel** [comment]
Also, where should I start looking at the implementation of this to understand it?

**Comment by nomeata** [comment]
TBH, I don’t think I will want to put much energy into maintaining this code, so I am less looking for contributors than for maintainers, if you know what I mean.

I would recommend reading first about the related work linked from README. For example, https://github.com/Chymyst/curryhoward says “When a type can be implemented in more than one way, heuristics ("least information loss") are used to prefer implementations that are more likely to satisfy algebraic laws”, which is a good start.

The core of the algorithm is implemented in https://github.com/nomeata/ghc-justdoit/blob/master/GHC/LJT.hs

**Comment by BebeSparkelSparkel** [comment]
How much potential do you see here? Have you concluded that anything that any correct interpretation has a trivial implementation?

**Comment by nomeata** [comment]
Sorry, not sure what you mean :-/

**Comment by BebeSparkelSparkel** [comment]
Just trying to find out if you see value in this or if you are trying to get out of it because you don’t see much potential

**Comment by nomeata** [comment]
As far as I know, nobody is using `ghc-justdoit` “for real”, so really, whatever people feel like hacking on is fine.

**Comment by weskerfoot** [comment]
I would love to see this as an editor or repl plugin that outputs the definition (or even gives you a list of possible ones). I'm not sure if this already exists with other tools like djinn.

**Comment by nomeata** [comment]
This tool works on GHC Core, so it does not produce code you can just “plug in” in Haskell Source. That would be a separate project (e.g. djinn).
